src/middleware/compression.rs

```diff
@@ -0,0 +1,434 @@
+pub use accept_encoding::Encoding;
+use async_compression::stream;
+use futures::future::BoxFuture;
+use http::{header::CONTENT_ENCODING, status::StatusCode, HeaderMap};
+use http_service::{Body, Request};
+
+use crate::{
+    middleware::{Middleware, Next},
+    response::IntoResponse,
+    Context, Error, Response,
+};
+
+/// Encode settings for the compression middleware.
+///
+/// This can be modified in the case that you want more control over the speed or quality of compression.
+///
+/// For more information on how to configure each of these settings, see the async-compression crate.
+#[derive(Debug)]
+pub struct EncodeSettings {
+    pub gzip: async_compression::flate2::Compression,
+    pub deflate: async_compression::flate2::Compression,
+    pub brotli: u32,
+    pub zstd: i32,
+}
+
+impl Default for EncodeSettings {
+    fn default() -> Self {
+        Self {
+            gzip: Default::default(),
+            deflate: Default::default(),
+            brotli: 11,
+            zstd: 3,
+        }
+    }
+}
+
+/// Middleware for automatically handling outgoing response compression.
+///
+/// This middleware currently supports HTTP compression using `gzip`, `deflate`, `br`, and `zstd`.
+#[derive(Debug)]
+pub struct Compression {
+    default_encoding: Encoding,
+    settings: EncodeSettings,
+}
+
+impl Default for Compression {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Compression {
+    /// Creates a new Compression middleware. The default encoding is [`Encoding::Identity`] (no encoding).
+    pub fn new() -> Self {
+        Self {
+            default_encoding: Encoding::Identity,
+            settings: Default::default(),
+        }
+    }
+
+    /// Creates a new Compression middleware with a provided default encoding.
+    ///
+    /// This encoding will be selected if the client has not set the `Accept-Encoding` header or `*` is set as the most preferred encoding.
+    pub fn with_default(default_encoding: Encoding) -> Self {
+        Self {
+            default_encoding,
+            settings: Default::default(),
+        }
+    }
+
+    /// Accesses a mutable handle to this middleware's [`EncodeSettings`].
+    ///
+    /// This will allow you to configure this middleware's settings.
+    pub fn settings_mut(&mut self) -> &mut EncodeSettings {
+        &mut self.settings
+    }
+
+    fn preferred_encoding(&self, headers: &HeaderMap) -> Result<Encoding, Error> {
+        let encoding = match accept_encoding::parse(headers) {
+            Ok(encoding) => encoding,
+            Err(e) => return Err(Error::from(StatusCode::BAD_REQUEST)),
+        };
+        match encoding {
+            Some(encoding) => Ok(encoding),
+            None => Ok(self.default_encoding),
+        }
+    }
+
+    /// Consumes the response and returns an encoded version of it.
+    fn encode(&self, mut res: Response, encoding: Encoding) -> Response {
+        // TODO: test and make sure this works
+        match encoding {
```

Comment:
There might be a way to eliminate duplicate codes... like, early-return on `Encoding::Identity` first, then make `match` evaluate to `Body`?

Reply:
(Now I think early-return might not be necessary)

Reply:
I think I could do this, but I'm not a *huge* fan of the way it looks (as I'd have to match the following `Encoding::Identity` arm with `unreachable!()`)... though I do agree, there is a lot of duplicate code here.

Reply:
I don't mind leaving as is, it was just a suggestion 🙂

Reply:
Oh, no problem :smile:

Yeah, it bugs me a little bit because if the client prefers no encoding (or our default encoding is set to `Identity`, then we're not supposed to touch the body at all (hence I can't do the swap until afterwards).

Reply:
@tirr-c I decided to go with reducing the boilerplate after all, it reads easier visually I think, especially with the new check in place for if a response was already encoded.